### ISSUE TITLE
MacOS Sierra Codesigning Command Failed

### ISSUE THREAD

**Issue by naderhen** [electron-userland/electron-builder#1486]
* **Version**: 17.0.1

electron-updater: 1.14.0

* **Target**: MacOS Sierra

I'm experiencing issues codesigning on Sierra. I've found Apple's documentation about correctly managing keychains in a CI environment to be rather confusing so I'm 99% sure this is a configuration issue on my end. If possible, I'd love your input if that is the case. Otherwise, this is the stack tract I experience with logging enabled:

```
[15:08:30][Step 14/16] Building OS X distribution.
[15:08:31][Step 14/16] Thu, 20 Apr 2017 15:08:31 GMT electron-builder Two package.json structure is used (dev: /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/package.json, app: /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/electron/package.json)
[15:08:31][Step 14/16] Thu, 20 Apr 2017 15:08:31 GMT electron-builder Effective config: {
[15:08:31][Step 14/16]   "appId": "com.electron.myapp",
[15:08:31][Step 14/16]   "publish": {
[15:08:31][Step 14/16]     "repo": null,
[15:08:31][Step 14/16]     "provider": "bintray",
[15:08:31][Step 14/16]     "user": "myname@mycompany",
[15:08:31][Step 14/16]     "package": "myapp",
[15:08:31][Step 14/16]     "owner": "mycompany"
[15:08:31][Step 14/16]   },
[15:08:31][Step 14/16]   "directories": {
[15:08:31][Step 14/16]     "app": "electron",
[15:08:31][Step 14/16]     "output": "release"
[15:08:31][Step 14/16]   },
[15:08:31][Step 14/16]   "icon": "/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/build/icon"
[15:08:31][Step 14/16] }
[15:08:31][Step 14/16] Installing app dependencies for arch x64 to /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/electron
[15:08:31][Step 14/16] Thu, 20 Apr 2017 15:08:31 GMT electron-builder Spawning /Users/teamcity/teamcity/system/jonnyzzz.nvm/v7.4.0/bin/node /Users/teamcity/teamcity/system/jonnyzzz.nvm/v7.4.0/lib/node_modules/npm/bin/npm-cli.js install --production --cache-min 999999999
[15:08:32][Step 14/16] myapp@0.21.0-418 /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/electron
[15:08:32][Step 14/16] `-- source-map-support@0.4.14 
[15:08:32][Step 14/16]   `-- source-map@0.5.6 
[15:08:32][Step 14/16] 
[15:08:32][Step 14/16] Thu, 20 Apr 2017 15:08:32 GMT electron-builder /Users/teamcity/teamcity/system/jonnyzzz.nvm/v7.4.0/bin/node (37670) exited with code 0
[15:08:32][Step 14/16] Packaging for darwin x64 using electron 1.4.15 to release/mac
[15:08:32][Step 14/16] Thu, 20 Apr 2017 15:08:32 GMT electron-builder Spawning /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/7zip-bin-mac/7za x -bd /Users/teamcity/Library/Caches/electron/electron-v1.4.15-darwin-x64.zip -o/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac
[15:08:32][Step 14/16] 
[15:08:32][Step 14/16] 7-Zip (a) [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
[15:08:32][Step 14/16] p7zip Version 16.02 (locale=utf8,Utf16=on,HugeFiles=on,64 bits,4 CPUs x64)
[15:08:32][Step 14/16] 
[15:08:32][Step 14/16] Scanning the drive for archives:
[15:08:32][Step 14/16] 1 file, 42120203 bytes (41 MiB)
[15:08:32][Step 14/16] 
[15:08:32][Step 14/16] Extracting archive: /Users/teamcity/Library/Caches/electron/electron-v1.4.15-darwin-x64.zip
[15:08:34][Step 14/16] --
[15:08:34][Step 14/16] Path = /Users/teamcity/Library/Caches/electron/electron-v1.4.15-darwin-x64.zip
[15:08:34][Step 14/16] Type = zip
[15:08:34][Step 14/16] Physical Size = 42120203
[15:08:34][Step 14/16] 
[15:08:34][Step 14/16] Everything is Ok
[15:08:34][Step 14/16] 
[15:08:34][Step 14/16] Folders: 143
[15:08:34][Step 14/16] Files: 180
[15:08:34][Step 14/16] Size:       112414322
[15:08:34][Step 14/16] Compressed: 42120203
[15:08:34][Step 14/16] Thu, 20 Apr 2017 15:08:34 GMT electron-builder /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/7zip-bin-mac/7za (37672) exited with code 0
[15:08:34][Step 14/16] Thu, 20 Apr 2017 15:08:34 GMT electron-builder Dev or extraneous dependencies: ../../electron-builder.yml, ../../electron-builder.json, ../../electron-builder.json5
[15:08:34][Step 14/16] Thu, 20 Apr 2017 15:08:34 GMT electron-builder Executing security find-identity -v
[15:08:34][Step 14/16] Thu, 20 Apr 2017 15:08:34 GMT electron-builder Executing security find-identity -v -p codesigning
[15:08:34][Step 14/16]   1) 35A5FDBE1C505C940FE0BBA78AA4299F521CC3C4 "Developer ID Application: My Name (GROUPNUMBER)"
[15:08:34][Step 14/16]   2) 35A5FDBE1C505C940FE0BBA78AA4299F521CC3C4 "Developer ID Application: My Name (GROUPNUMBER)"
[15:08:34][Step 14/16]   3) 35A5FDBE1C505C940FE0BBA78AA4299F521CC3C4 "Developer ID Application: My Name (GROUPNUMBER)"
[15:08:34][Step 14/16]      3 valid identities found
[15:08:34][Step 14/16] 
[15:08:34][Step 14/16]   1) 35A5FDBE1C505C940FE0BBA78AA4299F521CC3C4 "Developer ID Application: My Name (GROUPNUMBER)"
[15:08:34][Step 14/16]   2) B0D98E69AD404DB202320BEDA7CBC1FBBC58A568 "MyCompanys-Mac-mini.local"
[15:08:34][Step 14/16]   3) B0D98E69AD404DB202320BEDA7CBC1FBBC58A568 "MyCompanys-Mac-mini.local"
[15:08:34][Step 14/16]   4) 35A5FDBE1C505C940FE0BBA78AA4299F521CC3C4 "Developer ID Application: My Name (GROUPNUMBER)"
[15:08:34][Step 14/16]   5) 35A5FDBE1C505C940FE0BBA78AA4299F521CC3C4 "Developer ID Application: My Name (GROUPNUMBER)"
[15:08:34][Step 14/16]   6) B0D98E69AD404DB202320BEDA7CBC1FBBC58A568 "MyCompanys-Mac-mini.local"
[15:08:34][Step 14/16]   7) B0D98E69AD404DB202320BEDA7CBC1FBBC58A568 "MyCompanys-Mac-mini.local"
[15:08:34][Step 14/16]      7 valid identities found
[15:08:34][Step 14/16] 
[15:08:34][Step 14/16] Signing app (identity: Developer ID Application: My Name (GROUPNUMBER))
[15:08:34][Step 14/16] Thu, 20 Apr 2017 15:08:34 GMT electron-osx-sign `identity` passed in arguments.
[15:08:34][Step 14/16] Thu, 20 Apr 2017 15:08:34 GMT electron-osx-sign Executing... security find-identity -v
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Identity: 
[15:08:35][Step 14/16]  > Name: Developer ID Application: My Name (GROUPNUMBER)
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Identity: 
[15:08:35][Step 14/16]  > Name: Developer ID Application: My Name (GROUPNUMBER)
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Identity: 
[15:08:35][Step 14/16]  > Name: Developer ID Application: My Name (GROUPNUMBER)
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Pre-sign operation enabled for provisioning profile: 
[15:08:35][Step 14/16]  * Disable by setting `pre-embed-previsioning-profile` to `false`.
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Pre-sign operation enabled for entitlements automation with versions >= `1.1.1`: 
[15:08:35][Step 14/16]  * Disable by setting `pre-auto-entitlements` to `false`.
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign No `provisioning-profile` passed in arguments, will find in current working directory and in user library...
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign No provisioning profile found, will not embed profile in app contents.
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Signing application... 
[15:08:35][Step 14/16]  > Application: /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac/myapp.app 
[15:08:35][Step 14/16]  > Platform: darwin 
[15:08:35][Step 14/16]  > Entitlements: undefined 
[15:08:35][Step 14/16]  > Child entitlements: undefined 
[15:08:35][Step 14/16]  > Additional binaries: undefined 
[15:08:35][Step 14/16]  > Identity: Developer ID Application: My Name (GROUPNUMBER)
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Walking... /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac/myapp.app/Contents
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Signing... /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac/myapp.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Executing... codesign --sign Developer ID Application: My Name (GROUPNUMBER) --force /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac/myapp.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework
[15:08:35][Step 14/16] Thu, 20 Apr 2017 15:08:35 GMT electron-osx-sign Error executing file: 
[15:08:35][Step 14/16]  > Stdout:  
[15:08:35][Step 14/16]  > Stderr: 
[15:08:35][Step 14/16] { Error: Command failed: codesign --sign Developer ID Application: My Name (GROUPNUMBER) --force /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac/myapp.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework
[15:08:35][Step 14/16] 
[15:08:35][Step 14/16]     at ChildProcess.exithandler (child_process.js:211:12)
[15:08:35][Step 14/16]     at emitTwo (events.js:106:13)
[15:08:35][Step 14/16]     at ChildProcess.emit (events.js:191:7)
[15:08:35][Step 14/16]     at maybeClose (internal/child_process.js:885:16)
[15:08:35][Step 14/16]     at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
[15:08:35][Step 14/16] From previous event:
[15:08:35][Step 14/16]     at MacPackager.sign (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/out/macPackager.js:276:11)
[15:08:35][Step 14/16]     at /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/src/macPackager.ts:82:26
[15:08:35][Step 14/16] From previous event:
[15:08:35][Step 14/16]     at /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/src/macPackager.ts:82:10
[15:08:35][Step 14/16]     at Generator.next (<anonymous>)
[15:08:35][Step 14/16] From previous event:
[15:08:35][Step 14/16]     at MacPackager.pack (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/out/macPackager.js:181:11)
[15:08:35][Step 14/16]     at /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/src/packager.ts:257:24
[15:08:35][Step 14/16] From previous event:
[15:08:35][Step 14/16]     at Packager.doBuild (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/out/packager.js:379:11)
[15:08:35][Step 14/16]     at /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/src/packager.ts:170:52
[15:08:35][Step 14/16]     at Generator.next (<anonymous>)
[15:08:35][Step 14/16]     at runCallback (timers.js:651:20)
[15:08:35][Step 14/16]     at tryOnImmediate (timers.js:624:5)
[15:08:35][Step 14/16]     at processImmediate [as _immediateCallback] (timers.js:596:5)
[15:08:35][Step 14/16] From previous event:
[15:08:35][Step 14/16]     at Packager.build (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/out/packager.js:274:11)
[15:08:35][Step 14/16]     at Object.<anonymous> (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/src/builder.ts:198:33)
[15:08:35][Step 14/16]     at Generator.next (<anonymous>)
[15:08:35][Step 14/16] From previous event:
[15:08:35][Step 14/16]     at Object.build (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/node_modules/electron-builder/out/builder.js:70:21)
[15:08:35][Step 14/16]     at doBuild (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/deploy/build.js:55:37)
[15:08:35][Step 14/16]     at build (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/deploy/build.js:68:27)
[15:08:35][Step 14/16]     at go (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/deploy/prepackage.js:64:32)
[15:08:35][Step 14/16]     at Object.<anonymous> (/Users/teamcity/teamcity/work/c989685d470ade19/my-repo/deploy/prepackage.js:72:1)
[15:08:35][Step 14/16]     at Module._compile (module.js:571:32)
[15:08:35][Step 14/16]     at Object.Module._extensions..js (module.js:580:10)
[15:08:35][Step 14/16]     at Module.load (module.js:488:32)
[15:08:35][Step 14/16]     at tryModuleLoad (module.js:447:12)
[15:08:35][Step 14/16]     at Function.Module._load (module.js:439:3)
[15:08:35][Step 14/16]     at Module.runMain (module.js:605:10)
[15:08:35][Step 14/16]     at run (bootstrap_node.js:420:7)
[15:08:35][Step 14/16]     at startup (bootstrap_node.js:139:9)
[15:08:35][Step 14/16]     at bootstrap_node.js:535:3
[15:08:35][Step 14/16]   killed: false,
[15:08:35][Step 14/16]   code: null,
[15:08:35][Step 14/16]   signal: 'SIGSEGV',
[15:08:35][Step 14/16]   cmd: 'codesign --sign Developer ID Application: My Name (GROUPNUMBER) --force /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac/myapp.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework' }
[15:08:35][Step 14/16] 
[15:08:35][Step 14/16] npm ERR! Darwin 16.5.0
[15:08:35][Step 14/16] npm ERR! argv "/Users/teamcity/teamcity/system/jonnyzzz.nvm/v7.4.0/bin/node" "/Users/teamcity/teamcity/system/jonnyzzz.nvm/v7.4.0/bin/npm" "run" "prepackage" "--" "-t" "mac" "-r" "0.21.0-418"
[15:08:35][Step 14/16] npm ERR! node v7.4.0
[15:08:35][Step 14/16] npm ERR! npm  v4.0.5
[15:08:35][Step 14/16] npm ERR! code ELIFECYCLE
[15:08:35][Step 14/16] npm ERR! my-repo@0.21.0 prepackage: `node --harmony-async-await deploy/prepackage.js "-t" "mac" "-r" "0.21.0-418"`
[15:08:35][Step 14/16] npm ERR! Exit status 1
[15:08:35][Step 14/16] npm ERR! 
[15:08:35][Step 14/16] npm ERR! Failed at the my-repo@0.21.0 prepackage script 'node --harmony-async-await deploy/prepackage.js "-t" "mac" "-r" "0.21.0-418"'.
[15:08:35][Step 14/16] npm ERR! Make sure you have the latest version of node.js and npm installed.
[15:08:35][Step 14/16] npm ERR! If you do, this is most likely a problem with the my-repo package,
[15:08:35][Step 14/16] npm ERR! not with npm itself.
[15:08:35][Step 14/16] npm ERR! Tell the author that this fails on your system:
[15:08:35][Step 14/16] npm ERR!     node --harmony-async-await deploy/prepackage.js "-t" "mac" "-r" "0.21.0-418"
[15:08:35][Step 14/16] npm ERR! You can get information on how to open an issue for this project with:
[15:08:35][Step 14/16] npm ERR!     npm bugs my-repo
[15:08:35][Step 14/16] npm ERR! Or if that isn't available, you can get their info via:
[15:08:35][Step 14/16] npm ERR!     npm owner ls my-repo
[15:08:35][Step 14/16] npm ERR! There is likely additional logging output above.
[15:08:35][Step 14/16] 
[15:08:35][Step 14/16] npm ERR! Please include the following file with any support request:
[15:08:35][Step 14/16] npm ERR!     /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/npm-debug.log
[15:08:35][Step 14/16] Process exited with code 1

```

In order to make this work, I've done a few things.

1. Manually set `CSC_IDENTITY_AUTO_DISCOVERY=false`
2. Because of https://github.com/electron-userland/electron-builder/blob/e531017e58657257bab6cf8a92f596e50cf2ec2f/packages/electron-builder/src/codeSign.ts#L212  I assume it still considers my keychain != null and`isCi` is still true so I'm manually unsetting `CONTINUOUS_INTEGRATION`, `BUILD_NUMBER` and `TEAMCITY_VERSION`. At this point, it will finally respect my `CSC_IDENTITY_AUTO_DISCOVERY` setting. 
3. In my release script I am then using the `electron-osx-sign` library to `signAsync` and everything works out.

**Interesting Note**: If I SSH into my Mac Mini (used for CI) and attempt to run the codesigning command above I am able to reproduce the stacktrace:

```
codesign --sign Developer ID Application: My Name (GROUPNUMBER) --force /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac/myapp.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework
```

However, if I replace `Developer ID Application: My Name (GROUPNUMBER)` with `35A5FDBE1C505C940FE0BBA78AA4299F521CC3C4` it succeeds. Not sure if that's helpful,

Please let me know if there's anything else I can provide!

Best,
Nader

**Comment by naderhen** [comment]
Just because I've seen it mentioned in other threads, should say that I have neither `electron-rebuild` or `electron-prebuilt` in my dependencies or installed in my node_modules via something else

**Comment by blahah** [comment]
```
odesign --sign Developer ID Application: My Name (GROUPNUMBER) --force /Users/teamcity/teamcity/work/c989685d470ade19/my-repo/release/mac/myapp.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework
```

o/ this is not a valid shell command. It looks like it's probably an issue with how you're configuring the code signing.

Can you share the config you're using, but replace any sensitive data with random words?

**Comment by godza** [comment]
Is this solved?

**Comment by develar** [comment]
Fixed in 19.5.1.

**Comment by godza** [comment]
Thanks @develar
